### PR TITLE
feat: countgd sam2 video support

### PR DESCRIPTION
**Pull request by hrnn**
Added `countgd_sam2_video_tracking` tool, to use `countgd` for object detection and pass the results to `sam2` to track the objects in the entire video.

Screenshot 2024-12-05 161441
[Link to Colab](https://colab.research.google.com/drive/1GiuaaVGfBSSBUfzfiPz4OOA22n14XWti)

Note to the reviewer:
- I had to [refactor a bit my initial solution](https://github.com/landing-ai/vision-agent/pull/318/commits/c1320b8853aad4ae59628116f1b09779dbbf40b1) due to `mypy` complains when using too generic parameters.

### CODE REVIEW

File: tests/integ/test_tools.py [hunk 1]
```diff
@@ -40,6 +40,10 @@
     vit_image_classification,
     vit_nsfw_classification,
 )
+from vision_agent.tools.tools import (
+    countgd_sam2_video_tracking,
+    owlv2_sam2_video_tracking,
```
Comment (hrnn) [inline tests/integ/test_tools.py:45]: this is temporary, we'll remove it once we are sage to add it to `FUNCTION_TOOLS`

File: vision_agent/tools/tools.py [hunk 2]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
     return image
 
 
+class ODModels(str, Enum):
+    COUNTGD = "countgd"
+    OWLV2 = "owlv2"
+
+
+def od_sam2_video_tracking(
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2634]: nice

File: vision_agent/tools/tools.py [hunk 3]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
     return image
 
 
+class ODModels(str, Enum):
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2629]: Should we add in florence2 here as well?

File: vision_agent/tools/tools.py [hunk 4]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
...
+
+def countgd_sam2_video_tracking(
+    prompt: str,
+    frames: List[np.ndarray],
+    chunk_length: Optional[int] = 10,
+) -> List[List[Dict[str, Any]]]:
+    """'countgd_sam2_video_tracking' is a tool that can segment multiple objects given a text
+    prompt such as category names or referring expressions. The categories in the text
+    prompt are separated by commas. It returns a list of bounding boxes, label names,
+    mask file names and associated probability scores of 1.0.
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2725]: minro comment, only florence2 returns probability scores of 1.0, countgd and owlv2 will can return regular probability scores. So you can just say "and associated probability scores."

File: vision_agent/tools/tools.py [hunk 5]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
...
+def owlv2_sam2_video_tracking(
+    prompt: str,
+    frames: List[np.ndarray],
+    chunk_length: Optional[int] = 10,
+    fine_tune_id: Optional[str] = None,
+) -> List[List[Dict[str, Any]]]:
+    """'owlv2_sam2_video_tracking' is a tool that can segment multiple objects given a text
+    prompt such as category names or referring expressions. The categories in the text
+    prompt are separated by commas. It returns a list of bounding boxes, label names,
+    mask file names and associated probability scores of 1.0.
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2770]: see comment above on prob scores

File: vision_agent/tools/tools.py [hunk 6]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
...
+                'score': 1.0,
+                'label': 'dinosaur',
+                'bbox': [0.1, 0.11, 0.35, 0.4],
+                'mask': array([[0, 0, 0, ..., 0, 0, 0],
+                    [0, 0, 0, ..., 0, 0, 0],
+                    ...,
+                    [0, 0, 0, ..., 0, 0, 0],
+                    [0, 0, 0, ..., 0, 0, 0]], dtype=uint8),
+            },
+        ]
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2753]: Use the return values and examples from `florence2_sam2_video_tracking`. It's actually a list of list of dictionaries where the inner list is a frame

File: vision_agent/tools/tools.py [hunk 7]
```diff
@@ -2625,6 +2626,187 @@ def _plot_counting(
...
```
Comment (dillonalaird) [inline vision_agent/tools/tools.py:2798]: see comment above on return comments

**Review by dillonalaird — CHANGES_REQUESTED** [review]
looks good! just some minor comments

File: vision_agent/tools/tools.py (hunk 3 above)
Comment (hrnn) [reply vision_agent/tools/tools.py:2629]: Done

File: vision_agent/tools/tools.py (hunk 4 above)
Comment (hrnn) [reply vision_agent/tools/tools.py]: Fixed

File: vision_agent/tools/tools.py (hunk 5 above)
Comment (hrnn) [reply vision_agent/tools/tools.py]: Fixed

File: vision_agent/tools/tools.py (hunk 6 above)
Comment (hrnn) [reply vision_agent/tools/tools.py]: Fixed

File: vision_agent/tools/tools.py (hunk 7 above)
Comment (hrnn) [reply vision_agent/tools/tools.py]: Fixed

**Pull request merged**